File: styler-service/src/main/java/com/lookgen/styler/service/SketchScheduler.java
package com.lookgen.styler.service;

import com.lookgen.styler.domain.Photo;
import com.lookgen.styler.domain.Sketch;
import com.lookgen.styler.domain.Stage;
import com.lookgen.styler.domain.User;
import com.lookgen.styler.repo.PhotoRepository;
import com.lookgen.styler.repo.SketchRepository;
import com.lookgen.styler.repo.UserRepository;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;

@Service
public class SketchScheduler {

    private static final Logger log = LoggerFactory.getLogger(SketchScheduler.class);

    private final UserRepository userRepo;
    private final PhotoRepository photoRepo;
    private final SketchRepository sketchRepo;
    private final OpenAiClient openAi;
    private final MeterRegistry meterRegistry;

    public SketchScheduler(UserRepository userRepo, PhotoRepository photoRepo,
                           SketchRepository sketchRepo, OpenAiClient openAi,
                           MeterRegistry meterRegistry) {
        this.userRepo = userRepo;
        this.photoRepo = photoRepo;
        this.sketchRepo = sketchRepo;
        this.openAi = openAi;
        this.meterRegistry = meterRegistry;
    }

    @Scheduled(fixedDelayString = "${styler.poll-ms}")
    @Transactional
    public void tick() {
        List<User> users = userRepo.findPending(5);
        users.forEach(this::processUser);
    }

    private void processUser(User user) {
        Timer.Sample timer = Timer.start(meterRegistry);
        try {
            List<Photo> photos = photoRepo.findByUserIdOrderByOrd(user.getId());
            if (photos.size() != 5) {
                throw new IllegalStateException("Expected 5 photos");
            }
            List<String> urls = photos.stream()
                    .sorted(Comparator.comparing(Photo::getOrd))
                    .map(Photo::getUrl)
                    .toList();
            OpenAiClient.Response result = openAi.createSketch(urls, user.getStyle());
            Sketch sketch = new Sketch();
            sketch.setUserId(user.getId());
            sketch.setContent(result.getContent());
            sketch.setTokens(result.getTokens());
            sketchRepo.save(sketch);
            user.setStage(Stage.SKETCH_READY);
            user.setRetryCount((short) 0);
            meterRegistry.counter("sketch_jobs_total", "state", "done").increment();
        } catch (Exception ex) {
            log.warn("Failed to process user {}", user.getId(), ex);
            user.setRetryCount((short) (user.getRetryCount() + 1));
            if (user.getRetryCount() >= 3) {
                user.setStage(Stage.SKETCH_ERROR);
            }
            meterRegistry.counter("sketch_jobs_total", "state", "error").increment();
        } finally {
            user.setUpdatedAt(LocalDateTime.now());
            userRepo.save(user);
            timer.stop(meterRegistry.timer("sketch_latency_seconds"));
        }
    }
}

File: styler-service/src/main/kotlin/com/lookgen/styler/service/SketchScheduler.kt
package com.lookgen.styler.service

import com.lookgen.styler.config.StylerProperties
import com.lookgen.styler.domain.Stage
import com.lookgen.styler.domain.Sketch
import com.lookgen.styler.domain.User
import com.lookgen.styler.repo.PhotoRepository
import com.lookgen.styler.repo.SketchRepository
import com.lookgen.styler.repo.UserRepository
import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Timer
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
class SketchScheduler(
    private val userRepo: UserRepository,
    private val photoRepo: PhotoRepository,
    private val sketchRepo: SketchRepository,
    private val openAi: OpenAiClient,
    private val meterRegistry: MeterRegistry
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Scheduled(fixedDelayString = "\${styler.poll-ms}")
    @Transactional
    fun tick() {
        val users = userRepo.findPending(5)
        users.forEach { processUser(it) }
    }

    private fun processUser(user: User) {
        val timer = Timer.start(meterRegistry)
        try {
            val photos = photoRepo.findByUserIdOrderByOrd(user.id)
            if (photos.size != 5) throw IllegalStateException("Expected 5 photos")
            val urls = photos.sortedBy { it.ord }.map { it.url }
            val result = openAi.createSketch(urls, user.style)
            sketchRepo.save(Sketch(userId = user.id, content = result.content, tokens = result.tokens))
            user.stage = Stage.SKETCH_READY
            user.retryCount = 0
            meterRegistry.counter("sketch_jobs_total", "state", "done").increment()
        } catch (ex: Exception) {
            log.warn("Failed to process user ${'$'}{user.id}", ex)
            user.retryCount = (user.retryCount + 1).toShort()
            if (user.retryCount >= 3) user.stage = Stage.SKETCH_ERROR
            meterRegistry.counter("sketch_jobs_total", "state", "error").increment()
        } finally {
            user.updatedAt = LocalDateTime.now()
            userRepo.save(user)
            timer.stop(meterRegistry.timer("sketch_latency_seconds"))
        }
    }
}
